### ISSUE TITLE
1817: Illegal route allowed to go through tokened D7 (update: can trick the game to allow you to go through tokened cities)

### ISSUE THREAD

**Issue by SLOFila** [tobymao/18xx#3476]
Hey guys, here's a bug I just found.

In 1817, I was able to run a route through a tokened city.

https://18xx.games/game/22209?action=1711

How to reproduce the bug:

I have not done extensive testing HOWEVER I have noticed that you need to follow a specific pattern to make the illegal route work. (Using the 8 train)
What I have done is the following:
This is how I created the route: D7->C8->B5->F3->H3->G6->J7 and THEN I went and clicked on D1. In doing so, it seems that the game does not check the validity of the route in D7.
If I try to create the route by going a more natural way of D1->D7->C8 then the game immediately notifies me the route is illegal. Likewise if I try to go C8->D7->D1

Browser: Opera 73.0.3856.344 (desktop)
![illegalroute](https://user-images.githubusercontent.com/37855160/105607312-e07e3100-5d52-11eb-8a23-4405208ff49a.jpg)

**Comment by jenf** [comment]
Confirmed, very odd

**Comment by SLOFila** [comment]
I did some more testing.

I tried a similar scenario but instead of going to D1, I tried going to G6 (same clicking pattern, with G6 last). See the attached picture. It doesn't work. It recognizes it as an illegal route.

However, using the exact same route as in the picture, if I remove G6 and click on D1 then the route works again.

I tried something similar with the J company and the A20 red hex (trying to make it go through C22 using the same clicking pattern) and it correctly says the route is illegal.

![illegalroute2](https://user-images.githubusercontent.com/37855160/105612419-f7338080-5d70-11eb-8b3c-6ec4f01a54cc.jpg)

**Comment by SLOFila** [comment]
Further testing:
I went like this: D1->D7->C8->B5. As you can see in the attached picture, it says illegal route. However if I add one more stop, so instead of 4 steps it goes to 5 stops, for example from B5->F3 (or any other one) it says the route is legal.

I tried this with both the 6 train and the 8 train, same results.


![illegalroute3](https://user-images.githubusercontent.com/37855160/105612674-4af29980-5d72-11eb-8e8e-abf9a3dbee7c.jpg)

**Comment by SLOFila** [comment]
Even more testing and I think I found the problem:

I have cloned the game to try out different scenarios and I think this is the problem. If you look in the attached picture, I am running the J company and I am running the following route: C26-B17-D19-E22-C22-A20
If you look, C22 to A20 is illegal as J does not have a token at C22. What I believe happens is that when a route overlaps a previous legal route, for some reason the validity check is not triggered.

This issue might pop up in other games too that have 1817 style hexes (basically where the routes don't have a directional flow).

![illegalroute4](https://user-images.githubusercontent.com/37855160/105615566-7bdcc980-5d86-11eb-8772-663e6aec584c.jpg)
